compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/PropertyReferenceLowering.kt

```diff
@@ -0,0 +1,350 @@
+
+    private val IrMemberAccessExpression.setter: IrFunctionSymbol?
+        get() = (this as? IrPropertyReference)?.setter ?: (this as? IrLocalDelegatedPropertyReference)?.setter
+
+    private val IrMemberAccessExpression.field: IrFieldSymbol?
+        get() = (this as? IrPropertyReference)?.field
+
+    // Reflection metadata for local properties is serialized under the signature "<v#$N>" attached to the containing class.
+    // This maps properties to values of N.
+    private val localPropertyIndices = mutableMapOf<IrSymbol, Int>()
+
+    private val IrMemberAccessExpression.signature: String
+        get() = localPropertyIndices[getter]?.let { "<v#$it>" }
+            ?: getter?.owner?.let { context.state.typeMapper.mapSignatureSkipGeneric(it.descriptor).toString() }
+            // Plain Java fields do not have a getter, but can be referenced nonetheless.
+            ?: TODO("plain Java field signature")
+
+    private val IrMemberAccessExpression.symbol: IrSymbol
+        get() = getter?.owner?.symbol ?: field!!.owner.symbol
+
+    private val plainJavaClass =
+        context.getIrClass(FqName("java.lang.Class")).owner
+
+    private val reflectionClass =
+        context.getIrClass(FqName("kotlin.jvm.internal.Reflection")).owner
+
+    private val getOrCreateKotlinClass =
+        reflectionClass.functions.single { it.name.asString() == "getOrCreateKotlinClass" && it.valueParameters.size == 1 }
+
+    private val getOrCreateKotlinPackage =
+        reflectionClass.functions.single { it.name.asString() == "getOrCreateKotlinPackage" && it.valueParameters.size == 2 }
+
+    private val arrayItemGetter =
+        context.ir.symbols.array.owner.functions.single { it.name.asString() == "get" }
+
+    private val kPropertyType =
+        context.getIrClass(FqName("kotlin.reflect.KProperty")).owner.defaultType
+
+    private val kPropertiesFieldType =
+        context.ir.symbols.array.createType(false, listOf(makeTypeProjection(kPropertyType, Variance.OUT_VARIANCE)))
+
+    // Return some declaration the parent of which is the containing class/package for the referenced property.
+    // This is a bit of a hack, as the reason for not returning the container itself is the `mapOwner` call below.
+    private val IrMemberAccessExpression.propertyContainerChild: IrDeclaration?
+        get() {
+            var current: IrDeclaration? = getter?.owner ?: field?.owner
+            while (current?.parent is IrFunction)
+                current = current.parent as IrFunction
+            return current
+        }
+
+    // TODO: remove code duplication with CallableReferenceLowering
+    private val IrMemberAccessExpression.parentJavaClassReference
+        get() = IrClassReferenceImpl(
+            startOffset, endOffset,
+            plainJavaClass.defaultType,
+            plainJavaClass.symbol,
+            CrIrType(context.state.typeMapper.mapOwner(propertyContainerChild!!.descriptor))
```

Comment:
mapImplementationOwner: $$delegatedProperties are always located in DefaultImpls, not in interface

Reply:
@pyos Even with mapImplementationOwner $$delegatedProperties are located and initialized in interface not interface DefaultImpl. Moving this phase after 'interfacePhase' should help, but could you add some clarification about java field problem?

Reply:
The problem is that for reflection to work correctly, getSignature should return the signature of a property's getter, even if it does not actually exist (because the property is a Java field or the getter has been removed by PropertiesToFieldsLowering). I don't see an obvious way to do that using the current type mapper.

Reply:
@pyos Oh, so could you add several TODOs: about location of $$delegatedProperties and it's visibility and I will merge pr

Reply:
@max-kammerer done, added the second TODO and a bunch of other comments.